---
output: html_document
---
trunc analysis
========================================================

```{r setup, echo=TRUE, message=FALSE, cache=FALSE}
knitr::opts_chunk$set(fig.width=4, fig.height=4, echo=FALSE, width = 300)
library(knitr)
library(ff)
library(ffbase)
library(VariantAnnotation)
library(RColorBrewer)
library(ggthemes)
setwd("/Users/snafu/Documents/Project/germVar")
load(file=".RData")
```

Overview
---------------------

Overview of the allele frequency of all the truncating variants. `r nrow(info(trunc_VARu))` variants in `r length(unique(info(trunc_VARu)$Gene))` 

```{r trunc_AC_hist, eval=TRUE, echo=FALSE, message=FALSE}
library(gridExtra)
to_plot <- merge(as.data.frame(info(trunc_VARu)), list_goi, by="Gene")
breaks <- c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 200)
breaks_x <- c(0, 1, 3, 5, 10,  25, 50, 100, 150, 250, 400, 650, 1000)
breaks_y <- c(0, 1, 2, 10, 25, 50, 100, 250, 500, 1000, 1500, 2000)
#breaks <- c(0, 1,  3, 5, 8, 13, 20, 35, 55, 90, 150, 250, 400, 650, 1000)
to_plot$AC_bin <- cut(to_plot$EAC, breaks = breaks)
#p1 <- ggplot(aes(AC_bin), data=to_plot) + theme_few() + stat_bin(right=F,  fill="grey80") 
#p1 <- p1 + theme(axis.text.x= element_text(hjust=0.5, angle=90), legend.position = c(.8, .8) ) 
#p1 <- p1 + xlab("AC")
#p1
p1 <- ggplot(aes(AC), data=to_plot) + theme_few() + stat_bin(right=F,  fill="grey80") + scale_x_sqrt(breaks=breaks_x) + scale_y_sqrt(breaks=breaks_y)
p1 <- p1 + theme(axis.text.x= element_text(hjust=0.5, angle=0), legend.position = c(.8, .8), axis.title=element_text(size=rel(.7)) ) 
p1 <- p1 + xlab("Allele count") + ylab("Number of variants")
p1
```

```{r trunc_var_per_gene_hist, eval=TRUE, echo=FALSE, fig.width=6, fig.height=5., message=FALSE}
#trunc_bygene <- merge(as.data.frame(table(subset(info(trunc_VARu), tier1)$Gene)), as.data.frame(table(subset(info(lof_X2kG), tier1)$Gene)), by="Var1", all.x=T) %>% 
#                arrange(., -Freq.x) %>% subset(., Freq.x>=5) %>% mutate(., Freq.y=replaZero(Freq.y)) %>% plyr::rename(., c("Var1"="Gene", "Freq.x"="TCGA", "Freq.y"="X2kG"))
trunc_bygene <- merge(as.data.frame(table(subset(info(trunc_VARu), tier1)$Gene)), as.data.frame(table(subset(info(lof_X2kG), tier1)$Gene)), by="Var1", all.x=T) %>% 
                arrange(., -Freq.x) %>%  mutate(., Freq.y=replaZero(Freq.y)) %>% plyr::rename(., c("Var1"="Gene", "Freq.x"="TCGA", "Freq.y"="X2kG"))
#trunc_bygene <- cbind(trunc_bygene, data.frame(t(mapply(varNumFtest, trunc_bygene$TCGA, trunc_bygene$X2kG)))) %>% subset(., pval<0.2)  %>% arrange(., pval)
trunc_bygene <-  trunc_bygene %>% arrange(., -TCGA)  
trunc_bygene$Gene <- factor(trunc_bygene$Gene, levels=trunc_bygene$Gene)

#source("http://www.r-statistics.com/wp-content/uploads/2012/01/source_https.r.txt")
#source_https("https://raw.github.com/talgalili/R-code-snippets/master/Barnard.R")
library(gridExtra)
p1 <- ggplot(aes( Gene, Freq, fill=Source), data=reshape2::melt(trunc_bygene[c("Gene", "TCGA", "X2kG")][1:30, ], id.vars = "Gene", value.name="Freq", variable.name="Source") )
p1 <- p1 + theme_few() + geom_bar(stat="identity", position="dodge")
p1 <- p1 +  theme(axis.text.x= element_text(vjust=0.5, angle=90, size=rel(0.7)), legend.position = c(.85, .8)) + scale_fill_manual(values=c("grey10", "grey80"), guide = guide_legend(title = ""), labels=c("TCGA", "1000G"))
p1 <- p1 + xlab("") + ylab("Variant count") + scale_y_continuous(limits=c(0, 45)) + theme(axis.title.y=element_text(size=rel(0.7)))
p1
ggsave(filename="trunc_var_per_gene_hist.png",height=4, width=4)
#trunc_bygene <- plyr::join(trunc_bygene, list_goi[c("Gene", "cgc500")], by="Gene")
#trunc_1 <- subset(trunc_bygene, Gene %in% subset(list_goi, cgc500)$Gene)
#trunc_2 <- subset(trunc_bygene, Gene %in% subset(list_goi, !cgc500)$Gene)
#trunc_bygene$X2kG_N_lof <- replaZero(trunc_bygene$X2kG_N_lof)
#p1 <- ggplot(aes( Gene, Freq, fill=Source), data=reshape2::melt(trunc_1[1:30,], id.vars = "Gene", value.name="Freq", variable.name="Source") )
#p1 <- p1 + theme_few() + geom_bar(stat="identity", position="dodge")
#p1 <- p1 +  theme(axis.text.x= element_text(vjust=0.5, angle=90, size=rel(0.7)), legend.position = c(.85, .8)) + scale_fill_manual(values=c("grey80", "grey10"), guide = guide_legend(title = ""))
#p1 <- p1 + xlab("") + ylab("Variant count") + scale_y_continuous(limits=c(0, 45)) + theme(plot.margin = unit(c(1, 1, 0, 1), "cm"))
#p1
#p2 <- ggplot(aes( Gene, Freq, fill=Source), data=reshape2::melt(trunc_2[1:30,], id.vars = "Gene", value.name="Freq", variable.name="Source") )
#p2 <- p2 + theme_few() + geom_bar(stat="identity", position="dodge")
#p2 <- p2 +  theme(axis.text.x= element_text(vjust=0.5, angle=90, size=rel(0.7)), legend.position = c(.85, .8)) + scale_fill_manual(values=c("grey80", "grey10"), guide = F)
#p2 <- p2 + xlab("") + ylab("Variant count") + scale_y_continuous(limits=c(0, 45)) + theme(plot.margin = unit(c(0, 1, 1, 1), "cm"))
#grid.draw(rbind(ggplotGrob(p1), ggplotGrob(p2), size="first"))

```

```{r trunc_cat_mosaic, eval=TRUE, echo=FALSE, message=FALSE}
#mosaic plot of CGC500 variants in TS vs ONCO
to_plot <- rbind(table(subset(list_goi, cgc500)$Cat), table(subset(merge(as.data.frame(info(trunc_VARu)), list_goi, by="Gene"), cgc500)$Cat))
rownames(to_plot) <- c("# Gene", "# Variant")
mosaicplot((to_plot), main="", color=c("white", "pink", "lightgreen", "lightblue"), cex.axis=1.5)
```

```{r trunc_mem_hist, eval=TRUE, echo=FALSE, message=FALSE}
to_plot <- merge(as.data.frame(subset(info(trunc_VARu), tier1)), list_goi, by="Gene")
to_plot$Cat <- factor(to_plot$Cat, levels=c("TSG", "OG", "DUAL", "OTHER" ))
p <- ggplot(aes(mem), data= to_plot) + theme_few()
p <- p + geom_bar(fill = "grey80" )
p <- p + geom_bar(aes(fill = Cat), data = subset(to_plot, cgc500))
p <- p +scale_fill_brewer(palette="Set2", guide = guide_legend(title = "", label.theme = element_text(angle=0)), labels=c( "TSG", "OG", "Dual", "Unkown")) 
p <- p +  xlab("Gene Source") + ylab("Variant count") + theme(legend.position=c(0.8, 0.8))
p

```

```{r trunc_TS_study_, eval=TRUE, echo=FALSE, message=FALSE, fig.width=7, fig.height=10}
TS_muts <-  joint_tally$MUT %>% merge(., as.data.frame(info(trunc_VARu))[c("var_uid", "Gene", "tier1")], by="var_uid") %>%
                    merge(., all_tcga[!duplicated(all_tcga$Patient),][c("Patient", "age", "disease")], by="Patient") %>%
                    merge(., list_goi[c("Gene", "cgc500", "Cat", "HER", "rep", "rep_id")]) %>%
                    subset(., HER) 
TS_muts$rep_id[is.na(TS_muts$rep_id)] <- "NONE" 

gene_disease_tally <- TS_muts %>% group_by(., rep_id, Gene, disease) %>% dplyr::summarise(., N_pat = length(unique(Patient[PIT]))) %>% 
                      merge(., subset(all_tcga, !duplicated(Patient)) %>% group_by(disease) %>% dplyr::summarise(study_size = length(unique(Patient))), by="disease") %>%
                      mutate(., P_pat = N_pat/study_size)
rep_disease_tally <- TS_muts  %>% group_by(., rep_id, disease) %>% dplyr::summarise(., N_pat = length(unique(Patient[PIT])))
gene_disease_tally$disease <- factor(gene_disease_tally$disease, hclust_factor(rep_disease_tally, "disease", "rep_id", "N_pat", F))
gene_disease_tally$Gene <- factor(gene_disease_tally$Gene, hclust_factor(gene_disease_tally, "Gene", "disease", "P_pat", F))
gene_disease_tally <- subset(gene_disease_tally, N_pat>0)

# 2D grid plot all HER LoF variants
p1 <- ggplot(aes(disease, Gene, label=N_pat), data=gene_disease_tally) + geom_point(aes(size=sqrt(N_pat)), shape=1, fill="white", color="darkgrey")
p1 <- p1 + scale_size_continuous(range = c(2, 8))
p1 <- p1 + theme( panel.background = element_rect(fill='white',colour='black'))
p1 <- p1 + facet_grid(rep_id~., drop=TRUE, scales="free_y", space = "free_y")
p1 <- p1 + theme( legend.position = "none")
p1 <- p1 + theme( panel.grid.major.y = element_line(linetype=3, color="darkgray"),  panel.grid.major.x = element_line(linetype=1, color="lightgray"))
p1 <- p1 + theme(axis.text.x= element_text(hjust=1., vjust = 0.5, angle=90)) + xlab("") +ylab("") 
p1 <- p1 + geom_text(data=subset(gene_disease_tally, N_pat>1), hjust=0.5, color="red", size=3)
p1 <- p1 + theme(plot.margin = unit(c(0,1,1,1), "cm"))
p1 <- p1 + theme(strip.text.y = element_text(angle=0))
#p1

disease_tally <- TS_muts %>% group_by(., disease) %>% dplyr::summarise(., N_pat = length(unique(Patient[PIT]))) %>%
                 merge(., subset(all_tcga, !duplicated(Patient)) %>% group_by(disease) %>% dplyr::summarise(study_size = length(unique(Patient))), by="disease") %>%
                 mutate(., P_pat = N_pat/study_size)
disease_tally$disease <- factor(disease_tally$disease, levels=levels(gene_disease_tally$disease))
p2 <- ggplot(disease_tally, aes(disease, P_pat, label=N_pat)) + geom_bar(stat="identity", fill="grey80", alpha=1) + scale_y_continuous(limits=c(0, 0.2))
p2 <- p2 + theme_few() + theme( axis.text.y = element_text(angle=90, hjust=0.5),  axis.text.x = element_blank(), axis.ticks.x=element_blank() ) 
p2 <- p2 + xlab("") + ylab("Percentage of\n patients affected")
p2 <- p2 + geom_text(vjust= 0) 
p2 <- p2 + theme(plot.margin = unit(c(1,2.2,0,2), "cm"))
#p2

grid.newpage()
pushViewport(viewport(layout=grid.layout(2, 1, heights= unit(c(0.2, 0.8), "null"))))
print(p2, vp=viewport(layout.pos.row=1,layout.pos.col=1))
print(p1, vp=viewport(layout.pos.row=2,layout.pos.col=1))

#grid.draw(p2, p1, ncol=1)
#grid.draw(rbind(ggplotGrob(p1), ggplotGrob(p2)))

#ages <- merge(subset(all_tcga, !duplicated(Patient))[c("Patient", "disease")], all_clin[c("Patient", "age")])
#ages$disease <- factor(ages$disease, levels=tally_disease$disease)
#p2 <- ggplot(aes(disease, age), data = subset(ages, Patient %in% to_plot$Patient)) + geom_boxplot(outlier.shape = NA)
#p2 <- p2 + geom_jitter( aes(color = disease), width=0.03, alpha=1, size=2, data= subset(ages, Patient %in% to_plot$Patient))
#p2 <- p2 + theme_few() + theme( axis.text.x = element_text(angle=90), legend.position="") + xlab("") + ylab("age")
#p2 <- p2 + scale_color_manual(values=colorRampPalette(brewer.pal(8, "Set1"))(length(unique(to_plot$disease))))
#p2
#pvals <- sapply( unique(to_plot$disease),  function(x){
#  print(x);
#  to_test <- subset(to_plot, disease==x);
#  bg <- subset(ages[c("Patient", "disease", "age")], disease==x)
#  return(oneWayTestCallSet(to_test, "age", bg))
#}, USE.NAMES=T)
#pvals[pvals<0.1]
```

```{r trunc_LOH_per_gene, eval=TRUE,  echo=FALSE, message=FALSE, fig.width=6, fig.height=6}
MUTs <- merge(joint_tally$MUT, as.data.frame(info(trunc_VARu))[c("var_uid", "tier1", "Gene")], by="var_uid")
MUTs <- subset(MUTs, PIT & tier1)
LOH_tally <- MUTs %>% group_by(., Gene) %>% dplyr::summarise(., N_mut = length(unique(mut_uid)), per_LOH = length(unique(mut_uid[LOH]))/length(unique(mut_uid)))
p1 <- ggplot(aes(N_mut, per_LOH, label=Gene), data = LOH_tally) + geom_jitter(position=position_jitter(width=0.1, height=0.02), size=4, alpha=0.2)
p1 <- p1 + theme_few() + xlab("Number of mutations per Gene") + ylab("Percentage of LOH")

p2 <- ggplot(aes(N_mut, per_LOH, label=Gene), data = subset(LOH_tally, N_mut>=10 & per_LOH>=0.2)) + geom_jitter(position=position_jitter(width=0.1, height=0.02), size=4, alpha=0.2)
p2 <- p2 + theme_few()  + scale_x_sqrt() + geom_text(data=subset(LOH_tally, per_LOH>=0.2 & N_mut>=10), position=position_jitter(width=0.2, height=0.05), size=4, alpha=0.8)
p2 <- p2 + xlab("") +ylab("") + theme(plot.margin = unit(c(0,0,0,0), "inch"))

grid.newpage()
vpb_ <- viewport(width = 1, height = 1, x = 0.5, y = 0.5)  # the larger map
vpa_ <- viewport(width = 0.7, height = 0.7, x = 0.6, y = 0.6)  # the inset in upper right
print(p1, vp = vpb_)
print(p2, vp = vpa_)
```

```{r trunc_cgc_mrnaz, eval=TRUE, echo=FALSE, message=FALSE, fig.width=5, fig.height=12}
lf_cgc_muts <- droplevels(merge(lf_muts, cgc_test[c("Gene", "sig", "padj")], by="Gene"))
lf_cgc_muts$Gene2 <- apply(lf_cgc_muts[c("Gene", "sig")], 1, function(x) paste0(x, collapse=""))
mrnaz <- lapply(cgc_test$Gene, getRNASeq)
mrnaz <- do.call(rbind, mrnaz)
lf_cgc_muts <- merge(lf_cgc_muts, mrnaz[c("Gene", "Patient", "normalized_count", "study", "mrnaz")], by=c("Gene", "Patient"))
# cap abs value to 4 
lf_cgc_muts$mrnaz[lf_cgc_muts$mrnaz>4] <- 4 
lf_cgc_muts$HER[lf_cgc_muts$Gene=="RAD51B"] <- T

p <- ggplot(aes(mrnaz, reorder(Gene2, mrnaz, median), color=Gene2), data=lf_cgc_muts ) + geom_jitter(alpha=0.8) #+ theme_few() 
p <- p + theme( panel.background = element_rect(fill='white',colour='grey'))
p <- p + facet_grid(HER~., scales="free_y", space="free") + scale_color_manual(values=colorRampPalette(brewer.pal(8, "Set1"))(length(unique(lf_cgc_muts$Gene))), guide="none")
p <- p + xlab("Tumor mRNA level(zscore)") + ylab("") + geom_vline(xintercept=0, color="red", linetype=1)
p <- p + theme( panel.grid.major.y = element_line(linetype=3, color="darkgray"), panel.grid.major.x = element_blank())
p <- p + theme(strip.background = element_rect(fill="white"))
p <- p + strip.text
p
ggsave(filename="trunc_cgc_mrnaz.png",height=7, width=4)



