trunc analysis
========================================================

```{r setup, echo=TRUE, results='hide', message=FALSE, cache=FALSE}
knitr::opts_chunk$set(fig.width=6, fig.height=5, echo=FALSE, width = 300)
library(knitr)
library(ff)
library(ffbase)
library(VariantAnnotation)
library(RColorBrewer)
library(ggthemes)
setwd("/Users/snafu/Documents/Project/germVar")
load(file=".RData")
```

Overview
---------------------

Overview of the allele frequency of all the nsSNP variants. 

```{r trunc_AC_hist, eval=TRUE, echo=FALSE, fig.width=9, fig.height=4., message=FALSE}
library(gridExtra)
to_plot <- merge(as.data.frame(info(trunc_VARu)), list_goi, by="Gene")
breaks <- c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 150)
#breaks <- c(0, 1,  3, 5, 8, 13, 20, 35, 55, 90, 150, 250, 400, 650, 1000)
to_plot$AC_bin <- cut(to_plot$EAC, breaks = breaks)
#p1 <- ggplot(aes(AC_bin, fill=mem), data=to_plot) + theme_few() + stat_bin(right=F) 
p1 <- ggplot(aes(AC_bin), data=to_plot) + theme_few() + stat_bin(right=F) 
p1 <- p1 + theme(axis.text.x= element_text(hjust=0.5, angle=90), legend.position = c(.8, .8) ) +  scale_fill_brewer(palette="Set2")
p1 <- p1 + xlab("AC")

by_gene <- as.data.frame(table(to_plot$Gene))
by_gene <- merge(by_gene, list_goi, by.x="Var1", by.y="Gene")
#p2 <- ggplot(aes(Freq, fill=mem), data=by_gene) + theme_few() + stat_bin(right=F) 
p2 <- ggplot(aes(Freq), data=by_gene) + theme_few() + stat_bin(right=F) 
p2 <- p2 + theme(axis.text= element_text(hjust=0.5), legend.position = c(.8, .8) ) +  scale_fill_brewer(palette="Set2")
p2 <- p2 + xlab("#Variant per Gene")
grid.draw(cbind(ggplotGrob(p1), ggplotGrob(p2), size="first"))
```




